proxyclient/hv.py

```diff
@@ -39,6 +39,19 @@ class HOOK(IntEnum):
     "data" / Hex(Int64ul),
 )
 
+class RangedDict(dict):
```

Comment:
I like the interface here (I was thinking of building something like this), but this does a linear search over the keys which is O(n), and it also doesn't handle overlapping ranges. Can we do this in O(log n) and handle arbitrary overlaps?

The algorithm I'm thinking about is like this: keep the regular dict stuff for the main storage, but then add a cache of *non-overlapping* intervals stored as a sorted list of (start, end, set(ranges)) tuples, initialized to [(0, 1<<64, set())]. On dict insertion, find the range of intervals that overlap the new range (if any) using the bisect module, and add the current range to all, splitting any that need to be split. Then lookup becomes a simple bisection of the interval set (O(log n)), and automatically gives you a set of matching ranges (you just need to check the interval end address to accept or reject the candidate range). Insertion would be O(log n) in the total interval count and O(n) in the number of overlapping intervals. If you want to support removal, you can just remove a range from the covered intervals; I don't think we need to bother with re-coalescing.

(There may be better ways of doing this, feel free to come up with something better :))

Reply:
Also, this will be very useful later, as I plan to have "MMIO watcher" classes that can handle specific peripherals, and they need to be able to register themselves as watching certain MMIO ranges; I could use this same data structure for that, mapping ranges to watcher callbacks.

Reply:
Also, since we'd be handling mutations and multiple results and such, at this point I'd drop the dict subclassing and just do it as a new class; all we really need are `.add(range, value)` and `.lookup(addr)` to begin with.

Reply:
replaced with `bisect` and sorted lists. address lookup is factor 30 faster. I also got rid of the dict as underlying storage. Storing range, device_id tuples works as well and is c. 10% faster.